pkg/services/ngalert/api/authorization.go

```diff
@@ -162,3 +171,78 @@ func (api *API) authorize(method, path string) web.Handler {
 
 	panic(fmt.Sprintf("no authorization handler for method [%s] of endpoint [%s]", method, path))
 }
+
+// GetDatasourceScopesFromAlertRule extracts data source scopes from an alert rule
+func getEvaluatorForAlertRule(rule *ngmodels.AlertRule) ac.Evaluator {
+	scopes := make([]ac.Evaluator, 0, len(rule.Data))
+	for _, query := range rule.Data {
+		if query.QueryType == expr.DatasourceType || query.DatasourceUID == expr.OldDatasourceUID {
+			continue
+		}
+		scopes = append(scopes, ac.EvalPermission(datasources.ActionQuery, dashboards.ScopeFoldersProvider.GetResourceScopeUID(query.DatasourceUID)))
+	}
+	return ac.EvalAll(scopes...)
+}
+
+// authorizeRuleChanges analyzes changes in the rule group, determines what actions the user is trying to perform and check whether those actions are authorized.
+// If the user is not authorized to perform the changes the function returns ErrAuthorization with description of what action is not authorized. If evaluator function returns error, the function returns it
+func authorizeRuleChanges(namespace *models.Folder, changes *changes, evaluator func(evaluator ac.Evaluator) bool) error {
+	namespaceScope := dashboards.ScopeFoldersProvider.GetResourceScope(strconv.FormatInt(namespace.Id, 10))
+	if len(changes.Delete) > 0 {
+		allowed := evaluator(ac.EvalPermission(ac.ActionAlertingRuleDelete, namespaceScope))
+		if !allowed {
+			return fmt.Errorf("%w user cannot delete alert rules that belong to folder %s", ErrAuthorization, namespace.Title)
+		}
+	}
+
+	var addAuthorized, updateAuthorized bool
+
+	if len(changes.New) > 0 {
+		addAuthorized = evaluator(ac.EvalPermission(ac.ActionAlertingRuleCreate, namespaceScope))
+		if !addAuthorized {
+			return fmt.Errorf("%w user cannot create alert rules in the folder %s", ErrAuthorization, namespace.Title)
+		}
+		for _, rule := range changes.New {
+			dsAllowed := evaluator(getEvaluatorForAlertRule(rule))
+			if !dsAllowed {
+				return fmt.Errorf("%w to create a new alert rule '%s' because the user does not have read permissions for one or many datasources the rule uses", ErrAuthorization, rule.Title)
+			}
+		}
+	}
+
+	for _, rule := range changes.Update {
+		// If data is not changed, skip authorization. Effectively, this give users
+		// that do not have access to query data sources used by the rule
+		// limited access to some of the rule's settings like the interval, for, labels, annotations.
```

Comment:
I'm not entirely sure about this change. In my head, this sounds conflicting. Access to a datasource should have maximum priority over any alerting permissions.

If you don't have access to the datasource, you shouldn't even be seeing the rule/instance in the first place. For the sake of moving forward, let's continue as is but I'd love if could get a second opinion. 

I'll follow up offline.

Reply:
I have updated the code to require access to data source in order to update any property of the alert rule. In a follow-up PR I will add this restriction to read rules